test/general/tizen_device_test.dart

```diff
@@ -106,6 +110,75 @@ void main() {
     expect(processManager, hasNoRemainingExpectations);
   });
 
+  testWithoutContext(
+      'TizenDevice.installApp uninstalls and reinstalls TPK if installation fails',
+      () async {
+    final TizenDevice device = _createTizenDevice(
+      processManager: processManager,
+      fileSystem: fileSystem,
+      logger: logger,
+    );
+    final TizenTpk tpk = TizenTpk(
+      applicationPackage: fileSystem.file('app.tpk')..createSync(),
+      manifest: _FakeTizenManifest(),
+    );
+
+    processManager.addCommands(<FakeCommand>[
+      FakeCommand(
+        command: _sdbCommand(<String>['capability']),
+        stdout: <String>[
+          'cpu_arch:armv7',
+          'secure_protocol:disabled',
+          'platform_version:4.0',
+        ].join('\n'),
+      ),
+      FakeCommand(
+        command: _sdbCommand(<String>['shell', 'app_launcher', '-l']),
+        stdout: '''
+Application List for user 5001
+User's Application
+  Name    AppID
+=================================================
+'TestApp'     'TestApplication'
+''',
+      ),
+      FakeCommand(
+        command: _sdbCommand(<String>[
+          'pull',
+          '/opt/usr/apps/TestPackage/author-signature.xml',
+          '/.tmp_rand0/rand0/author-signature.xml',
+        ]),
+        onRun: () {
+          fileSystem
+              .file('/.tmp_rand0/rand0/author-signature.xml')
+              .createSync(recursive: true);
+        },
+      ),
+      FakeCommand(
+        command: _sdbCommand(<String>['install', 'app.tpk']),
+        stdout: '''
+__return_cb req_id[1] pkg_type[tpk] pkgid[TestPackage] key[install_percent] val[25]
+__return_cb req_id[1] pkg_type[tpk] pkgid[TestPackage] key[error] val[-11]
+__return_cb req_id[1] pkg_type[tpk] pkgid[TestPackage] key[end] val[fail]
+processing result : Author certificate not match [-11] failed
+''',
+      ),
+      FakeCommand(
+        command: _sdbCommand(<String>['uninstall', 'TestPackage']),
+        stdout: '''
+__return_cb req_id[1] pkg_type[tpk] pkgid[TestPackage] key[install_percent] val[100]
+__return_cb req_id[1] pkg_type[tpk] pkgid[TestPackage] key[end] val[ok]
+''',
+      ),
+      FakeCommand(command: _sdbCommand(<String>['install', 'app.tpk'])),
+    ]);
+
+    expect(await device.installApp(tpk), isTrue);
```

Comment:
How about setting the value of `apiVersion` in `_FakeTizenManifest`?
You can increase the coverage of the `installApp()` function by one more line.

[tizen_device.dart]
```
     260                 :          2 :     final Version? platformVersion = Version.parse(_platformVersion);
     261                 :          3 :     final Version? apiVersion = Version.parse(app.manifest.apiVersion);
     262                 :            :     if (platformVersion != null &&
     263                 :            :         apiVersion != null &&
     264                 :          0 :         apiVersion > platformVersion) {   // <---- here
     265                 :          0 :       _logger.printWarning(
```

Reply:
Thanks. I added a test for the version comparison.